File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    
    
    
    int trapFresh(vector<int> &height)
    {
        int lmax = INT_MIN;
        int rmax = INT_MIN;
        
        int n = height.size();
        int totalWater = 0;
        
        int i = 0;
        int j = n-1; 
        
        while(i<=j)
        {
            lmax = max(lmax, height[i]);
            rmax = max(rmax, height[j]);
            if(lmax < rmax)
            {
                totalWater += lmax - height[i];
                i++;
            }
            else
            {
                totalWater += rmax - height[j];
                j--;
            }
        }
        return totalWater;
    }
    
    
    
    int trap(vector<int>& height) 
    {
        int lmax = 0;
        int rmax = 0;
        int i = 0;
        int j = height.size() - 1;
        int volume = 0;
        while( i <= j ) 
        {
            lmax = max(height[i], lmax);
            rmax = max(height[j], rmax);
            if( lmax < rmax )
            {
                volume += lmax - height[i++];
            }
            else if ( rmax < lmax)
            {
                volume += rmax - height[j--];
            }
            else
            {
                volume += lmax - height[i];
                if ( j != i ) {
                    volume += rmax - height[j];
                    j--;
                }
                i++;
            }
        }
        // return volume;
        return trapFresh(height);
        
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;

        int i = 0;
        int j = n-1;

        int lmax = height[0];
        int rmax = height[n-1];

        int trappedWater = 0;
        while ( i<=j ) {
            lmax = Math.max(height[i], lmax);
            rmax = Math.max(height[j], rmax);

            if (lmax < rmax) {
                trappedWater += lmax - height[i];
                i++;
            }
            else {
                trappedWater += rmax - height[j];
                j--;
            }
        }
        return trappedWater;
    }
}